boot/bootstate20.go

```diff
@@ -0,0 +1,339 @@
+	}
+
+	return sn, nil, false, nil
+}
+
+// TODO:UC20: implement this
+func (s20 *bootState20Base) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
+	return update, nil
+}
+
+// TODO:UC20: implement this with modeenv
+func (s20 *bootState20Base) setNext(nextKernel snap.PlaceInfo) (rebootRequired bool, u bootStateUpdate, err error) {
+	return false, nil, nil
+}
+
+type bootState20Kernel struct{}
+
+func newBootState20(typ snap.Type) bootState {
+	switch typ {
+	case snap.TypeKernel:
+		return &bootState20Kernel{}
+	case snap.TypeBase:
+		return &bootState20Base{}
+	default:
+		panic("unsupported snap type update")
+	}
+}
+
+func (s20 *bootState20Kernel) revisions() (theSnap, theTrySnap snap.PlaceInfo, trying bool, err error) {
+	var bootSnap, tryBootSnap snap.PlaceInfo
+
+	// the trying snap is the one given by the try-kernel.efi symlink, while
+	// the known-good snap is the one given by the kernel.efi symlink
+	// we can determine if we are trying the try snap by the presence of
+	// 1. a try-kernel.efi symlink
+	// 2. kernel_status == trying
+
+	bl, err := bootloader.Find("", nil)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot find any bootloader: %v", err)
+	}
+
+	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
+	if !ok {
+		return nil, nil, false, fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", ebl.Name())
+	}
+
+	// get the kernel for this bootloader
+	bootSnap, err = ebl.Kernel()
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot locate kernel snap with bootloader %s: %v", ebl.Name(), err)
+	}
+
+	tryKernel, tryKernelExists, err := ebl.TryKernel()
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot locate try kernel snap with bootloader %s: %v", ebl.Name(), err)
+	}
+
+	if tryKernelExists {
+		tryBootSnap = tryKernel
+	}
+
+	m, err := bl.GetBootVars("kernel_status")
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ebl.Name(), err)
+	}
+
+	trying = (m["kernel_status"] == "trying")
+
+	return bootSnap, tryBootSnap, trying, nil
+}
+
+func (s20 *bootState20Kernel) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
+	var u20 *bootStateUpdate20
+	var err error
+	u20, err = newBootStateUpdate20(update, snap.TypeKernel)
+	if err != nil {
+		return nil, err
+	}
+
+	// kernel_status goes from "" -> "try" -> "trying" -> ""
+	// so if we are not in "trying" mode, nothing to do here
+	if u20.env["kernel_status"] != "trying" {
+		return u20, nil
+	}
+
+	// if kernel_status is trying, then we need to figure out what kernel we did
+	// boot from
+	sn, exists, err := u20.ebl.TryKernel()
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		// somehow we booted with kernel_status=trying, but no try-kernel
+		// symlink :-/
+		return nil, fmt.Errorf("cannot mark boot successful: kernel_status is \"trying\", but no try-kernel enabled")
```

Comment:
would it make sense to log the current `u20.ebl.Kernel()` for diagnostic purpose too?

Reply:
yes that's not a bad idea and that kind of log would be very helpful to an operator I think

Reply:
I'd like to do this in a followup if that's alright

Reply:
A followup sounds fine.